library(shiny)
library(shinyFiles)
library("shinyjs")
library(shinythemes)
library(limma)

library(DNAcopy)
library(SNPchip)
library(pracma)
library(data.table)
library("htmltools")
library(clusterSim)
library(car)

# Plot Normalisation
plot_test = read.table("test_data.txt", header = F, sep = "\t")
colnames(plot_test) = c("R","G")

#source('directoryInput.R')

options(shiny.maxRequestSize = 1000*1024^2)

ui <- tagList( useShinyjs(),
               tags$head(tags$title("START-R")),
               tags$head(tags$link(href = "PCNA.ico", rel ="shortcut icon")),
      
               HTML('<link rel="stylesheet" type="text/css" href="style.css" />'),
  
               img(src = "Logo_START_R.svg", class = "logo"),
               
               navbarPage(img(src = "PCNA.png", class = "icon"),id = "Workflow",
                
                          
                #==================================================
                # Presentation
                #==================================================
                
                tabPanel(title = "Workflow", value = "pres", 
                         HTML('<progress value="0" max="100"></progress>'),
                         h1("Welcome", class= "center"),
                         p('DNA replication is a highly regulated process involved in 
                            maintaining of genome stability. It relies on a spatio-temporal program determining 
                            where and when replication starts along the genome. The replication timing (RT) 
                            program is finely tuned and reflects the high order organization of the metazoan 
                            genome. It is clear that this program evolves with cell differentiation and cell fate. 
                            Moreover aberrant DNA RT  has been reported in many diseases including cancer.', class="intro"),
                         p('The protocol developed to study the RT of a given cell type is well established. 
                            We have improved the method permitting analysis of RT of the entire genome 
                            in mammalian cells. We implemented two bioinformatic tools, START-R 
                            (Simple Tool for Analysis of Replication Timing with R) and START-R viewer. 
                            START-R written in R language, makes it possible to analyse the data from 
                            microarrays with five steps: normalization, smoothing, TTR (Timing Transition Region) 
                            and CTR (Constant Timing Region) detection and finally differential analysis. 
                            For each step, several options are available and selectable by the user. 
                            At the end, with a classical computer configuration, the analysis takes few minutes. 
                            START-R viewer is a tool that allows to visualize the data generated by 
                            START-R as does a genome browser. These softwares have a user-friendly 
                            interface requiring only a “simple click” and can be run under Windows, 
                            Mac OS X and Linux systems without virtual machine. The script will 
                            remain open to the scientific community so that the software can always be improved or developed.
                           ', class="intro"),
                         
                         HTML('<a href="http://www.ijm.fr/" class ="no_style"><button type="button" class="run">Visit website !</button></a>'),
                         
                         p('This web application will allow you to enter the necessary informations for 
                            the analysis to run correctly. All the steps are organized 
                            in workflow. Each input is checked to avoid incorrect analyses.
                           ', class="intro"),
                         
                         actionButton('NextS1', 'Next', class = 'intro')
                ),     
                #==================================================
                # Step 1 - General
                #==================================================
                
                 tabPanel(title = "Initialisation", value = "Step1",
                          HTML('<progress value="17" max="100"></progress>'),
                          HTML("<div class='CG'>"),
                          h1("Initialisation", class ="center"),
                          fluidRow(
                            column(6,
                              h3("Organism", class = "center"),
                              p("Two types of organisms can be studied in START-R", class = "center"),
                              HTML("<div class='center'>"),
                              radioButtons("organism", label = NA,
                                           choices = list("Human" = "Human", "Mouse" = "Mouse"),
                                           selected = "Human", inline = T),
                              HTML("</div>"),
                              uiOutput("orga_img")
                              
                            ),
                            column(6,
                              h3("Differential", class = "center"),
                              p("Detect differences between 2 experiments", class = "center"),
                              HTML("<div class='center'>"),
                              radioButtons("dif", label = NA,
                                           choices = list("Yes" = "Yes", "No" = "No"),
                                           selected = "No", inline = T),
                              HTML("</div>"),
                              img(src= "differential.png", alt = 'dif', class = "dif")
                              
                            )
                          ),
                          
                          actionButton('NextS2', 'Next'),
                          actionButton('PrevPres', 'Prev.'),
                          HTML("</div>")
                  ),
                
                #==================================================
                # Step 2 _ File select
                #==================================================
                 
                 tabPanel(title = "File select", value = "Step2",
                          HTML('<progress value="34" max="100"></progress>'),
                          
                          h1("Select file"),
                          p("In this step, you must choose your files to analyze."),
                          
                          h2("Experiment 1"),
                          fluidRow(
                            
                            column(3,
                                   h3("Replica 1"),
                                   fileInput("file_E1_R1", label = NA),
                                   # Horizontal line ----
                                   tags$hr(),
                                   
                                   # Input: Checkbox if file has header ---- NOT CONNECTED
                                   radioButtons("header_E1_R1", "Header",
                                                choices = c("Yes" = TRUE,
                                                            "No" = FALSE),
                                                selected = TRUE, inline=T),
                                   
                                   # Input: Select separator ----
                                   radioButtons("sep_E1_R1", "Separator",
                                                choices = c(Comma = ",",
                                                            Semicolon = ";",
                                                            Tab = "\t"),
                                                selected = "\t", inline=T),
                                   
                                   # Input: Select quotes ----
                                   radioButtons("quote_E1_R1", "Quote",
                                                choices = c(None = "",
                                                            "Double Quote" = '"',
                                                            "Single Quote" = "'"),
                                                selected = "", inline=T),
                                   
                                   # Input: Select number of rows to skip ----
                                   p(tags$b("Number of microarray information lines to skip")),
                                   numericInput("skip_E1_R1", label = NA, value = 9,
                                                min = 0, max = 15),
                                   
                                   # Input: Select number of rows to display ----
                                   radioButtons("disp_E1_R1", "Display",
                                                choices = c(Head = "head",
                                                            "More (100 lines)" = "more",
                                                            No = "no"),
                                                selected = "head", inline=T)
                                   
                                   
                                  
                            ),
                            h3("Preview"),
                            column(dataTableOutput(outputId = "contents_E1_R1"), width = 9)
                          ),
                          
                          tags$hr(),
                          
                          #actionButton('reset_E1_R1', 'Reset Input'),
                          
                          fluidRow(
                            
                            column(3,
                                   h3("Replica 2"),
                                   fileInput("file_E1_R2", label = NA),
                                   # Horizontal line ----
                                   tags$hr(),
                                   
                                   # Input: Select number of rows to display ----
                                   radioButtons("disp_E1_R2", "Display",
                                                choices = c(Head = "head",
                                                            "More (100 lines)" = "more",
                                                            No = "no"),
                                                selected = "head", inline=T),
                                   
                                   # Input: Checkbox if file has header ---- NOT CONNECTED
                                   radioButtons("header_E1_R2", "Header",
                                                choices = c("Yes" = TRUE,
                                                            "No" = FALSE),
                                                selected = TRUE, inline=T),
                                   
                                   # Input: Select separator ----
                                   radioButtons("sep_E1_R2", "Separator",
                                                choices = c(Comma = ",",
                                                            Semicolon = ";",
                                                            Tab = "\t"),
                                                selected = "\t", inline=T),
                                   
                                   # Input: Select quotes ----
                                   radioButtons("quote_E1_R2", "Quote",
                                                choices = c(None = "",
                                                            "Double Quote" = '"',
                                                            "Single Quote" = "'"),
                                                selected = "", inline=T),
                                   
                                   # Input: Select number of rows to skip ----
                                   p(tags$b("Number of microarray information lines to skip")),
                                   numericInput("skip_E1_R2", label = NA, value = 9,
                                                min = 0, max = 15)

                            ),
                            h3("Preview"),
                            column(dataTableOutput(outputId = "contents_E1_R2"), width = 9)
                          ),
                          
                          h2("Experiment 2", id="title_E2"),
                          
                          fluidRow(id="row_E2_R1",
                            
                            column(3,
                                   h3("Replica 1"),
                                   fileInput("file_E2_R1", label = NA),
                                   # Horizontal line ----
                                   tags$hr(),
                                   
                                   # Input: Select number of rows to display ----
                                   radioButtons("disp_E2_R1", "Display",
                                                choices = c(Head = "head",
                                                            "More (100 lines)" = "more",
                                                            No = "no"),
                                                selected = "head", inline=T),
                                   
                                   # Input: Checkbox if file has header ---- NOT CONNECTED
                                   radioButtons("header_E2_R1", "Header",
                                                choices = c("Yes" = TRUE,
                                                            "No" = FALSE),
                                                selected = TRUE, inline=T),
                                   
                                   # Input: Select separator ----
                                   radioButtons("sep_E2_R1", "Separator",
                                                choices = c(Comma = ",",
                                                            Semicolon = ";",
                                                            Tab = "\t"),
                                                selected = "\t", inline=T),
                                   
                                   # Input: Select quotes ----
                                   radioButtons("quote_E2_R1", "Quote",
                                                choices = c(None = "",
                                                            "Double Quote" = '"',
                                                            "Single Quote" = "'"),
                                                selected = "", inline=T),
                                   
                                   # Input: Select number of rows to skip ----
                                   p(tags$b("Number of microarray information lines to skip")),
                                   numericInput("skip_E2_R1", label = NA, value = 9,
                                                min = 0, max = 15)

                            ),
                            h3("Preview"),
                            column(dataTableOutput(outputId = "contents_E2_R1"), width = 9)
                          ),
                          
                          tags$hr(),
                          
                          
                          fluidRow( id = "row_E2_R2",
                            
                            column(3,
                                   h3("Replica 2"),
                                   fileInput("file_E2_R2", label = NA),
                                   # Horizontal line ----
                                   tags$hr(),
                                   
                                   # Input: Select number of rows to display ----
                                   radioButtons("disp_E2_R2", "Display",
                                                choices = c(Head = "head",
                                                            "More (100 lines)" = "more",
                                                            No = "no"),
                                                selected = "head", inline=T),
                                   
                                   # Input: Checkbox if file has header ---- NOT CONNECTED
                                   radioButtons("header_E2_R2", "Header",
                                                choices = c("Yes" = TRUE,
                                                            "No" = FALSE),
                                                selected = TRUE, inline=T),
                                   
                                   # Input: Select separator ----
                                   radioButtons("sep_E2_R2", "Separator",
                                                choices = c(Comma = ",",
                                                            Semicolon = ";",
                                                            Tab = "\t"),
                                                selected = "\t", inline=T),
                                   
                                   # Input: Select quotes ----
                                   radioButtons("quote_E2_R2", "Quote",
                                                choices = c(None = "",
                                                            "Double Quote" = '"',
                                                            "Single Quote" = "'"),
                                                selected = "", inline=T),
                                   
                                   # Input: Select number of rows to skip ----
                                   p(tags$b("Number of microarray information lines to skip")),
                                   numericInput("skip_E2_R2", label = NA, value = 9,
                                                min = 0, max = 15)
                                   
                            ),
                            h3("Preview"),
                            column(dataTableOutput(outputId = "contents_E2_R2"), width = 9)
                          ),
                          
                          
                          fluidRow(
                            actionButton(inputId = "button", label = "Advanced file parameters", class = "no_float")
                          ),
                          
                          fluidRow(id = "Ad_file_box",
                                column(3,
                                textInput(inputId= "Green_signal", 
                                          label= "Column name of green signal :", 
                                          value = "gProcessedSignal")
                                ),
                                
                                column(3,
                                textInput(inputId= "Red_signal", 
                                          label= "Column name of red signal :", 
                                          value = "rProcessedSignal", placeholder = NULL)
                                ),
                                
                                column(3,
                                radioButtons("Late_frac", "Late fraction",
                                             choices = list("Cy3" = "Cy3", "Cy5" = "Cy5"
                                             ),
                                             selected = "Cy5")
                                ),
                                
                                column(3,
                                       radioButtons("Early_frac", "Early fraction",
                                                    choices = list("Cy3" = "Cy3", "Cy5" = "Cy5"
                                                    ),
                                                    selected = "Cy3")
                                )
                                
                            ),
                       
                          actionButton('NextS3', 'Next'),
                          actionButton('PrevS1', 'Prev.')
                          
                  ),
                
                #==================================================
                # Step 3 - Normalisation
                #==================================================
                 
                 tabPanel(title = "Normalisation", value = "Step3",
                          HTML('<progress value="51" max="100"></progress>'),
                          h1("Normalisation"),
                          p("In this step, you can choose the types of normalizations 
                            you want to apply to your data."),
                          
                          h2("Intra array"),
                          p("Short description"),
                          fluidRow(
                            column(3,radioButtons("RBintra_array", h4("Methods"),
                                                  choices = list("loess" = "loess", "control" = "control",
                                                                 "composite" = "composite", 
                                                                 "printtiploess" = "printtiploess",
                                                                 "median" = "median", "none" = "none", 
                                                                 "robustspline" = "robustspline"
                                                  ),
                                                  selected = "loess")),
                            column(3,
                                   h4("Description"),
                                   textOutput("Intra_array_description")
                            ),
                            column(6,
                                   plotOutput("scatterPlot_intra_array")
                            )
                          ),
                          
                          h2("Inter replica"),
                          p("Short description"),
                          fluidRow(
                            column(3,radioButtons("RBinter_replica", h4("Methods"),
                                                  choices = list("scale"= "scale",  
                                                                 "quantile" = "quantile",
                                                                 "none" = "none", "cyclicloess" = "cyclicloess" ), 
                                                  selected = "scale")),
                            column(3,
                                   h4("Description"),
                                   textOutput("Inter_replica_description")
                            ),
                            column(6,
                                   plotOutput("scatterPlot_inter_replica")
                            )
                          ),
                          
                          h2("Inter experience"),
                          p("Short description"),
                          fluidRow(
                            column(3,radioButtons("RBinter_experience", h4("Methods"),
                                                  choices = list("standardization" = "n1", 
                                                                 "without normalization" = "n0",
                                                                 "unitization" = "n3"), 
                                                 selected = "n1")),
                            column(3,
                                   h4("Description"),
                                   textOutput("Inter_experience_description")
                            ),
                            column(6,
                                   plotOutput("scatterPlot_inter_experience")
                            )
                          ),
                          
                          
                          column(12,h4("Cumulative normalisation visalisation"),
                                 p("Cumulative visualization is an example of the sequence
                                   of normalizations as performed in START-R."),
                                 plotOutput("scatterPlot_cumulative")
                          ),
                          
                          
                          actionButton('NextS4', 'Next'),
                          actionButton('PrevS2', 'Prev.')
                  
                 ),
                 
                #==================================================
                # Slide 4 - Analysis
                #==================================================
                
                tabPanel(title = "Analysis ", value = "Step4", 
                         HTML('<progress value="67" max="100"></progress>'),
                         
                         fluidRow(column(1,h4("Steps")),
                                  column(3, h4("Objective")),
                                  column(2, h4("Method")),
                                  column(3,h4("Method description")),
                                  column(3,h4("Additional options"))
                         ),
                         
                         
                         fluidRow(
                           column(1,checkboxInput("CH_Smooth", label = tags$b("Smooth"), value = TRUE)),
                           
                           column(3, p("Smooth description and objectives")),
                           
                           column(2,selectInput("select_method_smmoth", label = NA,
                                                choices = list("Loess" = "Loess", 
                                                              "Simple" = "s",
                                                               "Triangular" = "t", 
                                                               "Weighted" = "w",
                                                               "Modified" = "m", 
                                                               "Exponential" = "e", 
                                                               "Running" = "r"
                                                ), 
                                                selected = "loess")
                           ),
                           column(3,
                              p(textOutput("Smooth_method_description"))
                           ),
                           
                           column(3,
                                  h5("Span :"),
                                  p("Explication of span"),
                                  sliderInput("span_slider", label = NA,
                                              min = 100000, max = 10000000,
                                              step = 100000,
                                              value = 500000),
                                  
                                  h5("Windows size :"),
                                  p("Explication of size"),
                                  sliderInput("size_slider", label = NA,
                                              min = 10, max = 1000,
                                              step = 10,
                                              value = 1)
                                 
                                  )
                           
                           ),
                         
                         fluidRow(
                           column(1,checkboxInput("CH_TTR", label = tags$b("TTR"), value = TRUE)),
                           
                           column(3,
                                  p("TTR description and objectives"))
                           
                         ),
                         fluidRow(
                           column(1,checkboxInput("CH_segmentation", label = tags$b("Segmentation"), value = TRUE)),
                           
                           column(3, p("Segmentation description and objectives")),
                           
                           column(2),
                           column(3),
                           
                           column(3,
                                  h5("Windows size :"),
                                  p("Explication of SD"),
                                  sliderInput("SD_slider", label = NA,
                                              min = 1, max = 10,
                                              step = 0.1,
                                              value = 2.5)
                                  
                           )
                           
                         ),
                         
                         fluidRow(id = 'analysis_fusion',
                           column(1,checkboxInput("CH_Fusion", label = tags$b("Fusion"), value = TRUE)),
                           
                           column(3,
                                  p("Combine result of TTR and CTR"))
                           
                         ),
                         
                         fluidRow(id = 'analysis_dif',
                           column(1,checkboxInput("CH_Differential", label = tags$b("Differential"), value = TRUE)),
                           
                           column(3, p("Differntial description and objectives")),
                           
                           column(2,selectInput("select_method_differential", label = NA,
                                                choices = list("Euclidean method" = "Euclidean method", 
                                                               "Mean method" = "Mean method",
                                                               "Segment method" = "Segment method"
                                                ), 
                                                selected = "Mean method")
                           ),
                           column(3,
                                  p(textOutput("Differential_description"))
                           ),
                           
                           
                           column(3,
                                  h5("P-value threshold :", id = "PVT1"),
                                  p("Explication of P-value threshold", id = "PVT2"),
                                  numericInput("num_PVT", label = NA, value = 0.05,
                                               min = 0, max = 1),
                                  
                                  h5("Adjusted P-value :", id = "APV1"),
                                  p("Explication of Adjusted P-value", id = "APV2"),
                                  selectInput("select_method_ad_PV", label = NA,
                                              choices = list("holm" = "holm", 
                                                             "hochberg" = "hochberg", 
                                                             "hommel" = "hommel", 
                                                             "bonferroni" = "bonferroni", 
                                                             "BH" = "BH", 
                                                             "BY" = "BY",
                                                             "fdr" = "fdr", 
                                                             "none" = "none"
                                              ), 
                                              selected = "holm"),
                                  p(textOutput("ad_PV_description"), id = "APV3"),
                                  
                                  h5("Windows size :", id = "WS1"),
                                  p("Explication of Windows size", id = "WS2"),
                                  numericInput("num_WS", label = NA, value = 60,
                                               min = 10, max = 1000),
                                  
                                  h5("Overlap :", id = "Over1"),
                                  p("Explication of Overlap", id = "Over2"),
                                  numericInput("num_Over", label = NA, value = 30,
                                               min = 5, max = 500),
                                  
                                  h5("New parameter :", id = "NP1"),
                                  p("Explication of New parameter",  id = "NP2"),
                                  numericInput("num_NP", label = NA, value = 0.45,
                                               min = 0, max = 1)
                                  
                           )
                         ),
                                      
                         actionButton('NextS5', 'Next'),
                         actionButton('PrevS3', 'Prev.')
                
                ),
                #==================================================
                # Slide 5 - Outputs
                #==================================================
                
                tabPanel(title = "Outputs", value = "Step5", 
                         HTML('<progress value="84" max="100"></progress>'),
                         h1("Outputs"),
                         
                         # h3('Folder output'),
                         # p("Folder where analyses will be recorded"),
                         # shinyDirButton("dir", "Chose directory", "Upload"),
                         
                         #shinyDirButton('directory', 'Folder select', 'Please select a folder', class = "Select_folder"),
                         #textOutput("Select_dir"),
                         
                         radioButtons("file_outputs", h3("File Outputs (positions,intensity,...)"),
                                      choices = list(".bed" = "bed", ".txt" = "txt", "both" = "both"),
                                      selected = "both"),
                         
                         radioButtons("graphical_outputs", h3("Graphical outputs"),
                                      choices = list("Yes" = "Yes", "No" = "No"),
                                      selected = "Yes"),
                          
                         actionButton('NextS6', 'Next'),
                         actionButton('PrevS4', 'Prev.')
                ), 
                #==================================================
                # Slide 6 - Summary
                #==================================================
                
                 tabPanel(title = "Summary", value = "Step6",
                          HTML('<progress value="100" max="100"></progress>'),
                          h1("Summary"),
                           
                          fluidRow(
                            column(12, h3("General informations"), 
                                   textOutput("Out_orga"), 
                                   textOutput("Out_dif") 
                            )),
                          fluidRow(
                            column(12, h3("Selected Files"), 
                                   h4("Experience 1"),
                                   h5("Replica 1"),
                                   textOutput("Out_file_E1_R1"),
                                   h5("Replica 2"),
                                   textOutput("Out_file_E1_R2"),
                                   h4("Experience 2"),
                                   h5("Replica 1"),
                                   textOutput("Out_file_E2_R1"),
                                   h5("Replica 2"),
                                   textOutput("Out_file_E2_R2")
                            )),
                          
                          fluidRow(column(12, h3("Normalisation"), 
                                   textOutput("Out_intra_array") ,
                                   textOutput("Out_inter_replica") ,
                                   textOutput("Out_inter_experience") 
                            )),
                          fluidRow( 
                            column(3,h3("Analysis"),
                                   h4("Smooth"),
                                   textOutput("Out_smooth") ,
                                   textOutput("Out_smooth_method") ,
                                   textOutput("Out_smooth_span") ,
                                   textOutput("Out_smooth_size")
                            ),
                            column(3,
                                   h4("TTR"),
                                   textOutput("Out_TTR")
                            ),
                            
                            column(3,
                                     h4("Segmentation"),
                                     textOutput("Out_segmentation") ,
                                     textOutput("Out_segmentation_SD")
                            ),
                            column(3,
                                     h4("Differential"),
                                     textOutput("Out_differential") ,
                                     textOutput("Out_differential_method") ,
                                     textOutput("Out_differential_PVT") ,
                                     textOutput("Out_differential_PVAD"),
                                     textOutput("Out_differential_NP"),
                                     textOutput("Out_differential_Over"),
                                     textOutput("Out_differential_WS")
                            )
                            ),
                          fluidRow(
                            column(12, h3("Outputs"), 
                                   textOutput("Out_outputs_folder") ,
                                   textOutput("Out_outputs_file") ,
                                   textOutput("Out_outputs_graphical")
                            )
                          ),
                          actionButton('Validate', 'Validate'),
                          actionButton('PrevS5', 'Prev.')
                ),
                
                #==================================================
                # Slide 6 - Processing/ENd
                #==================================================
                tabPanel(title = "Job", value = "Job",
                         
                         h1('Click to run analysis', id = "H1_Run", align = "center"),
                         actionButton('Run', 'Run !', class="run"),
                         
                         h1('Processing', id = "H1_processing", align = "center"),
                         img(src = "process.gif", class = "process", id = "img_processing"),
                         
                         
                         h1('The analysis was successful!', id = "End", align = "center"),
                         #img(src = "end.gif", class = "end", id = "img_end")
                         HTML('<div class="checkmark" id = "img_end">
                              <svg version="1.1" id="Layer_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
                              viewBox="0 0 161.2 161.2" enable-background="new 0 0 161.2 161.2" xml:space="preserve">
                              <path class="path" fill="none" stroke="#31B404" stroke-miterlimit="10" d="M425.9,52.1L425.9,52.1c-2.2-2.6-6-2.6-8.3-0.1l-42.7,46.2l-14.3-16.4
                              c-2.3-2.7-6.2-2.7-8.6-0.1c-1.9,2.1-2,5.6-0.1,7.7l17.6,20.3c0.2,0.3,0.4,0.6,0.6,0.9c1.8,2,4.4,2.5,6.6,1.4c0.7-0.3,1.4-0.8,2-1.5
                              c0.3-0.3,0.5-0.6,0.7-0.9l46.3-50.1C427.7,57.5,427.7,54.2,425.9,52.1z"/>
                              <circle class="path" fill="none" stroke="#31B404" stroke-width="4" stroke-miterlimit="10" cx="80.6" cy="80.6" r="62.1"/>
                              <polyline class="path" fill="none" stroke="#31B404" stroke-width="6" stroke-linecap="round" stroke-miterlimit="10" points="113,52.8 
                              74.1,108.4 48.2,86.4 "/>
                              
                              <circle class="spin" fill="none" stroke="#31B404" stroke-width="4" stroke-miterlimit="10" stroke-dasharray="12.2175,12.2175" cx="80.6" cy="80.6" r="73.9"/>
                              
                              </svg>
                              ')
                        
                         )
  )
)

###################################################################################################
###################################################################################################
###################################################################################################


server <- function(input, output, session) {
  
  #========================================
  # Select folder
  #========================================
  # 
  # volumes <- getVolumes()
  # shinyDirChoose(input, 'directory', roots=volumes, session=session)
  # 
  # path1 <- reactive({
  #   return(parseDirPath(volumes, input$directory))
  # })
  # 
  # 
  # output$Select_dir <- renderText({ 
  #   #paste("Output folder :", readDirectoryInput(session, 'directory'))
  #   paste("Output folder :", path1())
  # })
  
  # dir
  shinyDirChoose(input, 'dir', roots = c(home = '~'), filetypes = c('', 'txt'))
  dir <- reactive(input$dir)
  output$dir <- renderPrint(dir())
  
  # path
  path1 <- reactive({
    home <- normalizePath("~")
    file.path(home, paste(unlist(dir()$path[-1]), collapse = .Platform$file.sep))
  })
  
  # files
  output$files <- renderPrint(list.files(path()))
  
  #========================================
  # Hide tabset when page is init
  #========================================
  
  observe({
    hideTab(inputId = "Workflow", target = "Step1")
  })
  
  observe({
    hideTab(inputId = "Workflow", target = "Step2")
  })
  
  observe({
    hideTab(inputId = "Workflow", target = "Step3")
  })
  
  observe({
    hideTab(inputId = "Workflow", target = "Step4")
  })
  
  observe({
     hideTab(inputId = "Workflow", target = "Step5")
  })
  
  observe({
    hideTab(inputId = "Workflow", target = "Step6")
  })
  
  observe({
    hideTab(inputId = "Workflow", target = "Job")
  })
  
  observe({
    hideElement(id = "H1_processing")
  })
  
  observe({
    hideElement(id = "img_processing")
  })
  
  observe({
    hideElement(id = "End")
  })
  
  observe({
    hideElement(id = "img_end")
  })
  
  observe({
    hideElement(id = "Ad_file_box")
  })
  
  
  #========================================
  # Button Nav workflow
  #========================================
  
  #---------
  # Pres
  #---------
  observeEvent(input$NextS1, {
    # Show next page in navbar
    showTab(inputId = "Workflow", target = "Step1")
    
    # Move in this next page
    updateTabsetPanel(session, "Workflow",
                      selected = "Step1")
  })
  
  #---------
  # Step 1
  #---------
  observeEvent(input$NextS2, {
    # Show next page in navbar
    showTab(inputId = "Workflow", target = "Step2")
    
    # Move in this next page
    updateTabsetPanel(session, "Workflow",
                      selected = "Step2")
  })
  
  observeEvent(input$PrevPres, {
    updateTabsetPanel(session, "Workflow",
                      selected = "pres")
  })
  
  #---------
  # Step 2 
  #---------
  observeEvent(input$PrevS1, {
    updateTabsetPanel(session, "Workflow",
                      selected = "Step1")
  })
  
  observeEvent(input$NextS3, {
    # Show next page in navbar
    showTab(inputId = "Workflow", target = "Step3")
    
    # Move in this next page
    updateTabsetPanel(session, "Workflow",
                      selected = "Step3")
  })
  
  #---------
  # Step 3 
  #---------
  observeEvent(input$PrevS2, {
    updateTabsetPanel(session, "Workflow",
                      selected = "Step2")
  })
  
  observeEvent(input$NextS4, {
    # Show next page in navbar
    showTab(inputId = "Workflow", target = "Step4")
    
    # Move in this next page
    updateTabsetPanel(session, "Workflow",
                      selected = "Step4")
  })
  
  #---------
  # Step 4 
  #---------
  observeEvent(input$PrevS3, {
    updateTabsetPanel(session, "Workflow",
                      selected = "Step3")
  })
  
  observeEvent(input$NextS5, {
    # Show next page in navbar
    showTab(inputId = "Workflow", target = "Step5")
    
    # Move in this next page
    updateTabsetPanel(session, "Workflow",
                      selected = "Step5")
  })
  
  #---------
  # Step 5 
  #---------
  observeEvent(input$PrevS4, {
    updateTabsetPanel(session, "Workflow",
                      selected = "Step4")
  })
  
  observeEvent(input$NextS6, {
    # Show next page in navbar
    showTab(inputId = "Workflow", target = "Step6")
    
    # Move in this next page
    updateTabsetPanel(session, "Workflow",
                      selected = "Step6")
  })
  
  #---------
  # Step 6 
  #---------
  observeEvent(input$PrevS5, {
    updateTabsetPanel(session, "Workflow",
                      selected = "Step5")
  })
  
  observeEvent(input$Validate, {
    # Show next page in navbar
    showTab(inputId = "Workflow", target = "Job")
    
    # Move in this next page
    updateTabsetPanel(session, "Workflow",
                      selected = "Job")
  })
  
  #========================================
  # Init
  #========================================

  output$orga_img <- renderUI({
    img(src = paste0(input$organism, ".svg"), height = 100, class = "center")
  })
  
  #========================================
  # Read File
  #========================================
  
  # observeEvent(input$reset_E1_R1, {
  #   reset("file_E1_R1")
  # })
  
  output$contents_E1_R1 <-  renderDataTable({
    
    req(input$file_E1_R1)
    
    df <- read.csv(input$file_E1_R1$datapath,
                   header = as.logical(input$header_E1_R1),
                   sep = input$sep_E1_R1,
                   quote = input$quote_E1_R1,
                   nrows=100,
                   skip = input$skip_E1_R1
                   )
    
    if(input$disp_E1_R1 == "head") {
      return(head(df))
    }
    else if(input$disp_E1_R1 == "more") {
      return(df)
    }else{
      return(NA)
    }
    
  },  options = list(scrollX = TRUE))
  
  output$contents_E1_R2 <-  renderDataTable({
    
    req(input$file_E1_R2)
    
    df <- read.csv(input$file_E1_R2$datapath,
                   header = as.logical(input$header_E1_R2),
                   sep = input$sep_E1_R2,
                   quote = input$quote_E1_R2,
                   nrows=100,
                   skip = input$skip_E1_R2
    )
    
    if(input$disp_E1_R2 == "head") {
      return(head(df))
    }
    else if(input$disp_E1_R2 == "more") {
      return(df)
    }else{
      return(NA)
    }
    
  },  options = list(scrollX = TRUE))
  
  output$contents_E2_R1 <-  renderDataTable({
    
    req(input$file_E2_R1)
    
    df <- read.csv(input$file_E2_R1$datapath,
                   header = as.logical(input$header_E2_R1),
                   sep = input$sep_E2_R1,
                   quote = input$quote_E2_R1,
                   nrows=100,
                   skip = input$skip_E2_R1
    )
    
    if(input$disp_E2_R1 == "head") {
      return(head(df))
    }
    else if(input$disp_E2_R1 == "more") {
      return(df)
    }else{
      return(NA)
    }
    
  },  options = list(scrollX = TRUE))
  
  output$contents_E2_R2 <-  renderDataTable({
    
    req(input$file_E2_R2)
    
    df <- read.csv(input$file_E2_R2$datapath,
                   header = as.logical(input$header_E2_R2),
                   sep = input$sep_E2_R2,
                   quote = input$quote_E2_R2,
                   nrows=100,
                   skip = input$skip_E2_R2
    )
    
    if(input$disp_E2_R2 == "head") {
      return(head(df))
    }
    else if(input$disp_E2_R2 == "more") {
      return(df)
    }else{
      return(NA)
    }
    
  },  options = list(scrollX = TRUE))
 
  #========================================
  # Read File - Adv
  #======================================== 
  observeEvent(input$Late_frac, {
    if(input$Late_frac == "Cy3"){
      updateRadioButtons(session, "Early_frac", selected = "Cy5")
    }else if(input$Late_frac == "Cy5"){
      updateRadioButtons(session, "Early_frac",selected = "Cy3")
    }
  })
  
  observeEvent(input$Early_frac, {
    if(input$Early_frac == "Cy3"){
      updateRadioButtons(session, "Late_frac",  selected = "Cy5")
    }else if(input$Early_frac == "Cy5"){
      updateRadioButtons(session, "Late_frac", selected = "Cy3")
    }
  })
  
  #========================================
  # Read File - next
  #========================================
  
  observe({
    if(input$dif == "No") {
      if(is.null(input$file_E1_R1) && is.null(input$file_E1_R2)){
        disable("NextS3")
      }else{
        enable("NextS3")
      }
    }else if (input$dif == "Yes") {
      if(is.null(input$file_E2_R1) && is.null(input$file_E2_R2)){
        disable("NextS3")
      }else{
        enable("NextS3")
      }
    }

  })
  
  
  #---------------------------
  # Show/hide dif Read file
  #---------------------------
  
  observeEvent(input$dif, {
    if(input$dif == "Yes"){
        showElement(id = "row_E2_R1")
        showElement(id = "row_E2_R2")
        showElement(id = "title_E2")
        showElement(id = "analysis_dif")
        updateCheckboxInput(session, "CH_Smooth", value = TRUE)
        updateCheckboxInput(session, "CH_TTR", value = TRUE)
        updateCheckboxInput(session, "CH_segmentation", value = TRUE)
        updateCheckboxInput(session, "CH_Fusion", value = TRUE)
        disable("CH_Smooth")
        disable("CH_TTR")
        disable("CH_segmentation")
        disable("CH_Fusion")

    }else{
      hideElement(id = "row_E2_R1")
      hideElement(id = "row_E2_R2")
      hideElement(id = "title_E2")
      hideElement(id = "analysis_dif")
      enable("CH_Smooth")
      enable("CH_TTR")
      enable("CH_segmentation")
      enable("CH_Fusion")
    }
  })
  
  #---------------------------------------
  # Advanced file parameters
  #---------------------------------------
  observeEvent(input$button, {
    
    if(input$button %% 2 != 1){
      shinyjs::hide(id = "Ad_file_box")
    }else{
      shinyjs::show(id = "Ad_file_box")
    }
  })
  
  
  #========================================
  # Description - Normalisation
  #========================================
  
  #----------------------------------------
  # Description - Intra array
  #----------------------------------------
  output$Intra_array_description <- renderText({ 
    
    if(input$RBintra_array == "loess"){
      "Description of loess"
    } else if(input$RBintra_array == "control"){
      "Oshlack et al (2004) consider the special issues that arise when a large 
      proportion of probes are differentially expressed. They propose an improved 
      version of composite loess normalization, which is implemented in the 
      control method. This fits a global loess curve through a set 
      of control spots, such as a whole-library titration series, and applies 
      that curve to all the other spots."
    }else if(input$RBintra_array == "composite"){
      "Description of composite"
    }else if(input$RBintra_array == "printtiploess"){
      "Description of printtiploess"
    }else if(input$RBintra_array == "median"){
      "Method median subtracts the weighted median from the M-values for each array"
    }else if(input$RBintra_array == "none"){
      "Method none computes M-values and A-values but does no normalization"
    }else if(input$RBintra_array == "robustspline"){
      "Method robustspline normalizes the M-values for a single microarray using robustly fitted 
      regression splines and empirical Bayes shrinkage."
    }
  })
  
  output$scatterPlot_intra_array <- renderPlot({
    #MA.l = normalizeWithinArrays(plot_test, method=input$RBintra_array) 
    
    matplot(cbind(plot_test ), type = "l", 
            lwd = c(1,1,2,2), lty = c(3,3,1,1),
            col = c("royalblue", "red", "royalblue", "red"),
            ylab = "Intensity", xlab = "Position" ,
            main = "Visualisation")
  })
  
  #----------------------------------------
  # Description - Inter replica
  #----------------------------------------
  
  output$Inter_replica_description <- renderText({ 
    
    if(input$RBinter_replica == "scale"){
      "Scale (method='scale') scales the columns to have the same median."
    } else if(input$RBinter_replica == "quantile"){
      "Quantile normalization was originally proposed by Bolstad et al (2003) 
      for Affymetrix-style single-channel arrays. 
      Quantile normalization forces the entire empirical 
      distribution of each column to be identical."
    }else if(input$RBinter_replica == "none"){
      "No normalization is calculated."
    }else if(input$RBinter_replica == "cyclicloess"){
      "Cyclic loess normalization was originally proposed by Bolstad et al (2003) 
      for Affymetrix-style single-channel arrays. Cyclic loess normalization applies 
      loess normalization to all possible pairs of arrays, usually cycling through 
      all pairs several times. Cyclic loess is slower than quantile, but allows 
      probe-wise weights and is more robust to unbalanced differential expression."
    }
  })
  
  
  output$scatterPlot_inter_replica <- renderPlot({
    MA.q = normalizeBetweenArrays(plot_test, method=input$RBinter_replica)
   
    matplot(cbind(plot_test , MA.q), type = "l", 
            lwd = c(1,1,2,2), lty = c(3,3,1,1),
            col = c("royalblue", "red", "royalblue", "red"),
            ylab = "Intensity", xlab = "Position",
            main = "Visualisation")
  })
 
  
  #----------------------------------------
  # Description - Inter experience
  #----------------------------------------

  # data.Normalization
  output$Inter_experience_description <- renderText({ 
    
    if(input$RBinter_experience == "n1"){
      "(x-mean)/sd"
    } else if(input$RBinter_experience == "n0"){
      "Without normalization"
    }else if(input$RBinter_experience == "n3"){
      "(x-mean)/range"
    }
    })
  
  
  output$scatterPlot_inter_experience <- renderPlot({
    z1 <- data.Normalization(plot_test,type=input$RBinter_experience,
                             normalization="column")
    matplot(cbind(plot_test , z1), type = "l", 
            lwd = c(1,1,2,2), lty = c(3,3,1,1),
            col = c("royalblue", "red", "royalblue", "red"),
            ylab = "Intensity", xlab = "Position" ,
            main = "Visualisation")
  })
  
  #----------------------------------------
  # Description - Cumulative
  #----------------------------------------
  
  output$scatterPlot_cumulative <- renderPlot({
    
    MA.q = normalizeBetweenArrays(plot_test, method=input$RBinter_replica)
    
    z1 <- data.Normalization(MA.q,type=input$RBinter_experience,
                             normalization="column")
    
    matplot(cbind(plot_test , z1), type = "l", 
            lwd = c(1,1,2,2), lty = c(3,3,1,1),
            col = c("royalblue", "red", "royalblue", "red"),
            ylab = "Intensity", xlab = "Position",
            main = "Visualisation")
  })
  
  
  #========================================
  # Description - Smooth
  #========================================
  
  output$Smooth_method_description <- renderText({ 
  
    if(input$select_method_smmoth== "loess"){
      "Description of loess"
    } else if(input$select_method_smmoth == "s"){
      "It computes the simple moving average. n indicates the number of previous 
      data points used with the current data point when calculating the moving average"
    }else if(input$select_method_smmoth == "t"){
      "It computes the triangular moving average by calculating the first simple 
      moving average with window width of ceil(n+1)/2; then it calculates a second 
      simple moving average on the first moving average with the same window size."
    }else if(input$select_method_smmoth == "w"){
      "It calculates the weighted moving average by supplying weights for each 
      element in the moving window. Here the reduction of weights follows a linear trend."
    }else if(input$select_method_smmoth == "m"){
      "It calculates the modified moving average. The first modified moving average 
      is calculated like a simple moving average. Subsequent values are calculated 
      by adding the new value and subtracting the last average from the resulting sum."
    }else if(input$select_method_smmoth == "e"){
      "It computes the exponentially weighted moving average. The exponential 
      moving average is a weighted moving average that reduces influences by 
      applying more weight to recent data points () reduction factor 2/(n+1)."
    }else if(input$select_method_smmoth == "r"){
      "Ihis is an exponential moving average with a reduction factor of 1/n ["
    }
    })
  
  #========================================
  # Description - Differential
  #========================================
  
  output$Differential_description <- renderText({ 
    
    if(input$select_method_differential == "Euclidean method"){
      "Description of Euclidean method"
    } else if(input$select_method_differential == "Mean method"){
      "Description of Mean method"
    }else if(input$select_method_differential == "Segment method"){
      "Description of Segment method"
    }
  })
  
  #========================================
  # show / hide - differential
  #========================================
  

  observeEvent(input$select_method_differential, {
    if(input$select_method_differential == "Euclidean method"){
      showElement(id = "PVT1")
      showElement(id = "PVT2")
      showElement(id = "num_PVT")
      
      showElement(id = "APV1")
      showElement(id = "APV2")
      showElement(id = "APV3")
      
      showElement(id = "WS1")
      showElement(id = "WS2")
      showElement(id = "num_WS")
      
      showElement(id = "Over1")
      showElement(id = "Over2")
      showElement(id = "num_Over")
      
      showElement(id = "NP1")
      showElement(id = "NP2")
      showElement(id = "num_NP")
      
    }else if(input$select_method_differential == "Mean method"){
      showElement(id = "PVT1")
      showElement(id = "PVT2")
      showElement(id = "num_PVT")
      
      showElement(id = "APV1")
      showElement(id = "APV2")
      showElement(id = "APV3")
      
      showElement(id = "WS1")
      showElement(id = "WS2")
      showElement(id = "num_WS")
      
      showElement(id = "Over1")
      showElement(id = "Over2")
      showElement(id = "num_Over")
      
      hideElement(id = "NP1")
      hideElement(id = "NP2")
      hideElement(id = "num_NP")
      
    }else if (input$select_method_differential == "Segment method"){
      hideElement(id = "PVT1")
      hideElement(id = "PVT2")
      hideElement(id = "num_PVT")
      
      hideElement(id = "APV1")
      hideElement(id = "APV2")
      hideElement(id = "APV3")
      
      hideElement(id = "WS1")
      hideElement(id = "WS2")
      hideElement(id = "num_WS")
      
      hideElement(id = "Over1")
      hideElement(id = "Over2")
      hideElement(id = "num_Over")
      
      hideElement(id = "NP1")
      hideElement(id = "NP2")
      hideElement(id = "num_NP")
    }
  })
  
  #========================================
  # Description - Differential - adjusted pval
  #========================================
  
  output$ad_PV_description <- renderText({ 
    if(input$select_method_ad_PV == "holm"){
      "Description of holm"
    } else if(input$select_method_ad_PV == "hochberg"){
      "Description of hochberg"
    }else if(input$select_method_ad_PV == "hommel"){
      "Description of hommel"
    }else if(input$select_method_ad_PV == "bonferroni"){
      "Description of bonferroni"
    }else if(input$select_method_ad_PV == "BH"){
      "Description of BH"
    }else if(input$select_method_ad_PV == "BY"){
      "Description of BY"
    }else if(input$select_method_ad_PV == "fdr"){
      "Description of fdr"
    }else if(input$select_method_ad_PV == "none"){
      "Description of none"
    }
  })
  
  #========================================
  # Analysis check box
  #========================================

  observeEvent(input$CH_Smooth, {
    if(input$CH_Smooth == FALSE){
      updateCheckboxInput(session, "CH_TTR", value = FALSE)
      updateCheckboxInput(session, "CH_segmentation", value = FALSE)
      updateCheckboxInput(session, "CH_Fusion", value = FALSE)
    }
    
  })
  
  observeEvent(input$CH_TTR, {
    if(input$CH_TTR == TRUE){
      updateCheckboxInput(session, "CH_Smooth", value = TRUE)
    }else if(input$CH_TTR == FALSE){
      updateCheckboxInput(session, "CH_segmentation", value = FALSE)
      updateCheckboxInput(session, "CH_Fusion", value = FALSE)
    }
  })
  
  
  observeEvent(input$CH_segmentation, {
    if(input$CH_segmentation == TRUE){
      updateCheckboxInput(session, "CH_Smooth", value = TRUE)
      updateCheckboxInput(session, "CH_TTR", value = TRUE)
    }else if(input$CH_segmentation == FALSE){
      updateCheckboxInput(session, "CH_Fusion", value = FALSE)
    }
  })
  
  observeEvent(input$CH_Fusion, {
    if(input$CH_Fusion == TRUE){
      updateCheckboxInput(session, "CH_Smooth", value = TRUE)
      updateCheckboxInput(session, "CH_TTR", value = TRUE)
      updateCheckboxInput(session, "CH_segmentation", value = TRUE)
    }
  })
  
  observeEvent(input$dif, {
    if(input$dif == "No"){
      updateCheckboxInput(session, "CH_Differential", value = FALSE)  
    }else if(input$dif == "Yes"){
      updateCheckboxInput(session, "CH_Differential", value = TRUE)
    }
    
  })
  
  #========================================
  # Select folder - next
  #========================================
  
  # observe({
  #   #if(readDirectoryInput(session, 'directory') == ""){
  #   if(is.null(path1())) { 
  #     disable("NextS6")
  #   }else{
  #     enable("NextS6")
  #   }
  # })
  # 
  #========================================
  # Summary
  #========================================
  
  ## Step 1 
  output$Out_orga <- renderText({ 
    paste("Organism :", input$organism)
  })
  
  output$Out_dif <- renderText({ 
    paste("Differential :", input$dif)
  })
  
  ## Step 2
  output$Out_file_E1_R1 <- renderText({ 
    paste("File name :", input$file_E1_R1$name)
  })
  
  output$Out_file_E1_R2 <- renderText({ 
    paste("File name :", input$file_E1_R2$name)
  })
  
  output$Out_file_E2_R1 <- renderText({ 
    paste("File name :", input$file_E2_R1$name)
  })
  
  output$Out_file_E2_R2 <- renderText({ 
    paste("File name :", input$file_E2_R2$name)
  })
  
  ## Step 3 - normalisatio
  
  output$Out_intra_array <- renderText({ 
    paste("Intra array :", input$RBintra_array)
  })
  
  output$Out_inter_replica <- renderText({ 
    paste("Inter replica :", input$RBinter_replica)
  })
  
  output$Out_inter_experience <- renderText({ 
    paste("Inter experience :", input$RBinter_experience)
  })
  
  ## Step 4 - Analyse
  
  output$Out_smooth <- renderText({ 
    paste("Realiazed? :", input$CH_Smooth)
  })
  
  output$Out_smooth_method <- renderText({ 
    paste("Method :", input$select_method_smmoth)
  })
  
  output$Out_smooth_span <- renderText({ 
    paste("Span :", input$span_slider)
  })
  
  output$Out_smooth_size <- renderText({ 
    paste("Size :", input$size_slider)
  })
  
  output$Out_TTR <- renderText({ 
    paste("Realiazed? :", input$CH_TTR)
  })
  
  output$Out_segmentation <- renderText({ 
    paste("Realiazed? :", input$CH_segmentation)
  })
  
  output$Out_segmentation_SD <- renderText({ 
    paste("SD :", input$SD_slider)
  })
  
  output$Out_differential <- renderText({ 
    paste("Realiazed? :", input$CH_Differential)
  })
  
  output$Out_differential_method <- renderText({ 
    paste("Method :", input$select_method_differential)
  })
  
  output$Out_differential_PVT <- renderText({ 
    paste("PVT :", input$num_PVT)
  })
  
  output$Out_differential_PVAD <- renderText({ 
    paste("Method adjustment PV :", input$select_method_ad_PV)
  })
  
  output$Out_differential_WS <- renderText({ 
    paste("Windows size:", input$num_WS)
  })
  
  output$Out_differential_Over <- renderText({ 
    paste("Overlap:", input$num_Over)
  })
  
  output$Out_differential_NP <- renderText({ 
    paste("New parameter:", input$num_NP)
  })
  
  ## Step 5 - Outputs

  # output$Out_outputs_folder <- renderText({ 
  #   #paste("Output folder :", readDirectoryInput(session, 'directory'))
  #   paste("Output folder :", path1())
  # })
  
  output$Out_outputs_file <- renderText({ 
    paste("File format :", input$file_outputs)
  })
  
  output$Out_outputs_graphical <- renderText({ 
    paste("Graphical :", input$graphical_outputs)
  })
  
  
  #========================================
  # Processing
  #========================================
  
  click = FALSE
  observeEvent(input$Run , {
    # Before
    hideElement(id = "Run")
    hideElement(id = "H1_Run")
    showElement(id = "H1_processing")
    showElement(id = "img_processing")
    showElement(id = "Processing")
    
    # Variable for code
    if(is.null(input$file_E1_R1$datapath)){
      File1 =  input$file_E1_R2$datapath
    }else{
      File1 = input$file_E1_R1$datapath
    }
    
    if(is.null(input$file_E1_R2$datapath)){
      File2 = input$file_E1_R1$datapath
    }else{
      File2 = input$file_E1_R2$datapath
    }
    
    if(is.null(input$file_E2_R1$datapath)){
      File3 = input$file_E2_R2$datapath
    }else{
      File3 = input$file_E2_R1$datapath
    }
    
    if(is.null(input$file_E2_R2$datapath)){
      File4 = input$file_E2_R1$datapath
    }else{
      File4 = input$file_E2_R2$datapath
    }

    nomtotal = c(File1 = File1,
    File2 = File2,
    File3 = File3,
    File4 = File4)

    sortie_image = input$graphical_outputs
    bed = input$file_outputs
    nor1 = input$RBintra_array 
    nor2 = input$RBinter_replica 
    nor3 = input$RBinter_experience
    
    e1 = input$CH_Smooth
    e2 = input$CH_TTR
    e3 = input$CH_segmentation
    e4 = input$CH_Fusion
    e5 = input$CH_Differential
    
    v1 = input$span_slider
    v2 = input$SD_slider
    v3 = input$select_method_ad_PV
    v4 = input$select_method_smmoth
    v5 = input$num_WS
      
    fs1 = input$skip_E1_R1
    fs2 = input$Green_signal 
    fs3 = input$Red_signal 
    fs4 = input$Late_frac 
    fs5 = input$Early_frac
  
    pv1 = input$num_PVT
    pv2 = input$num_WS
    pv3 = input$select_method_ad_PV
    pv4 = input$num_Over
    
    type_dif = input$select_method_differential
    seuil = input$num_NP
    organisme = input$organism
    
    #directory = readDirectoryInput(session, 'directory')
    # directory = path1()
    directory = "./Outputs"
    
    # run
    source('R/STARTR.R', local = TRUE)
    
    # After
    hideElement(id = "H1_processing")
    hideElement(id = "img_processing")
    hideElement(id = "Processing")
    hideElement(id = "timer")
    
    showElement(id = "img_end")
    showElement(id = "End")
  })
  

  
  output$beginTime <- renderText({
    begin = Sys.time()
    paste0("(The analysis started at : ",format(Sys.time(), "%X"),")")
  })
  
  output$endTime <- renderText({
    
    time = difftime(Sys.time(), begin, units='secs')
    minutes = trunc(time/60)
    secondes = trunc(time - minutes * 60);
    
    paste("The analysis has taken : ", minutes, "mins.",
          secondes, 'secs' )
  })
  
  

  
}



shinyApp(ui, server)